src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/Renamer.cs

```diff
@@ -1,170 +1,187 @@
-                         IRoslynServices roslynServices,
-                         Project project,
-                         string oldFilePath,
-                         string newFilePath)
+                         IUserNotificationServices userNotificationServices,
+                         IRoslynServices roslynServices)
         {
+            _projectVsServices = projectVsServices;
+            _unconfiguredProjectTasksService = unconfiguredProjectTasksService;
             _workspace = workspace;
-            _threadingService = threadingService;
-            _userNotificationServices = userNotificationServices;
             _environmentOptions = environmentOptions;
+            _userNotificationServices = userNotificationServices;
             _roslynServices = roslynServices;
-            _project = project;
-            _newFilePath = newFilePath;
-            _oldFilePath = oldFilePath;
         }
 
-        public async void OnWorkspaceChangedAsync(object sender, WorkspaceChangeEventArgs args)
+        public void HandleRename(string oldFilePath, string newFilePath)
         {
-            Document oldDocument = (from d in _project.Documents where StringComparers.Paths.Equals(d.FilePath, _oldFilePath) select d).FirstOrDefault();
-
-            if (oldDocument == null)
-                return;
-
-            if (args.Kind == WorkspaceChangeKind.DocumentAdded && args.ProjectId == _project.Id)
-            {
-                Project project = (from p in args.NewSolution.Projects where p.Id.Equals(_project.Id) select p).FirstOrDefault();
-                Document addedDocument = (from d in project?.Documents where d.Id.Equals(args.DocumentId) select d).FirstOrDefault();
-
-                if (addedDocument != null && StringComparers.Paths.Equals(addedDocument.FilePath, _newFilePath))
-                {
-                    _docAdded = true;
-                }
-            }
-
-            if (args.Kind == WorkspaceChangeKind.DocumentRemoved && args.ProjectId == _project.Id && args.DocumentId == oldDocument.Id)
-            {
-                _docRemoved = true;
-            }
+            // We do not need to block the completion of HandleRename so we queue it using the threading service
+            // and ignore the result.
+            // NOTE: we queue work on JTF queue so VS shutdown can happen cleanly
+            _projectVsServices.ThreadingService.RunAndForget(
+                () => HandleRenameAsync(oldFilePath, newFilePath),
+                    unconfiguredProject: _projectVsServices.Project);
+        }
 
-            if (args.Kind == WorkspaceChangeKind.DocumentChanged && args.ProjectId == _project.Id)
+        internal async Task HandleRenameAsync(string oldFilePath, string newFilePath)
+        {
+            // Do not offer to rename types if the user changes the file extensions
+            if (!oldFilePath.EndsWith(Path.GetExtension(newFilePath), StringComparison.OrdinalIgnoreCase))
```

Comment:
Thinking out loud here, but given that this only triggers of a rename of a `<Compile />` item, do we care about file extensions? There probably isn't a scenario where it's relevant, but in theory if something can be renamed and is still considered a compile input after the rename, then we might want to process it...

Reply:
Since compile is computed automatically in sdk style projects this is less of an issue, but if users have set `EnableDefaultCompileItems` to false we would still want this behavior of not offering a rename if we change the file from `.cs` to `.vb`   

---
In reply to: [288347951](https://github.com/dotnet/project-system/pull/4876#discussion_r288347951) [](ancestors = 288347951)

Reply:
I believe the assumption is that if the user is renaming the file type then our language service can't be right because the language that the file is presenting as is wrong / needs to be changed.